exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java

```diff
@@ -470,43 +500,73 @@ public void eval() {
+                duration = duration.minusDays(1);
+            } else if (from.isBefore(to) && duration.compareTo(java.time.Duration.ZERO) < 0) {
+                // positive period and negative duration
+                period = java.time.Period.between(from.toLocalDate(), to.toLocalDate().minusDays(1));
+                duration = duration.plusDays(1);
+            } else {
+                period = java.time.Period.between(from.toLocalDate(), to.toLocalDate());
+            }
+
+            out.months = (int) period.toTotalMonths();
+            out.days = period.getDays();
+            out.milliseconds = (int) duration.toMillis();
         }
     }
 
     @FunctionTemplate(name = "age", scope = FunctionTemplate.FunctionScope.SIMPLE, nulls = NullHandling.NULL_IF_NULL)
     public static class AgeDate2Function implements DrillSimpleFunc {
         @Param DateHolder right;
-        @Workspace long queryStartDate;
+        @Workspace java.time.OffsetDateTime queryStartDate;
         @Output IntervalHolder out;
         @Inject ContextInformation contextInfo;
 
         @Override
         public void setup() {
             int timeZoneIndex = contextInfo.getRootFragmentTimeZone();
-            org.joda.time.DateTimeZone timeZone = org.joda.time.DateTimeZone.forID(org.apache.drill.exec.expr.fn.impl.DateUtility.getTimeZone(timeZoneIndex));
-            org.joda.time.DateTime now = new org.joda.time.DateTime(contextInfo.getQueryStartTime(), timeZone);
-            queryStartDate = (new org.joda.time.DateMidnight(now.getYear(), now.getMonthOfYear(), now.getDayOfMonth(), timeZone)).getMillis();
+            java.time.ZoneId zoneId = java.time.ZoneId.of(org.apache.drill.exec.expr.fn.impl.DateUtility.getTimeZone(timeZoneIndex));
+            java.time.ZonedDateTime dt = java.time.Instant.ofEpochMilli(contextInfo.getQueryStartTime()).atZone(zoneId);
+            queryStartDate = java.time.OffsetDateTime.of(dt.toLocalDate(), java.time.LocalTime.MIDNIGHT, java.time.ZoneOffset.UTC);
         }
 
         @Override
         public void eval() {
-            long diff = queryStartDate - right.value;
-            long days = diff / org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis;
-            out.months = (int) (days / org.apache.drill.exec.vector.DateUtilities.monthToStandardDays);
-            out.days = (int) (days % org.apache.drill.exec.vector.DateUtilities.monthToStandardDays);
-            out.milliseconds = (int) (diff % org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis);
+            java.time.OffsetDateTime dt = java.time.Instant.ofEpochMilli(right.value).atOffset(java.time.ZoneOffset.UTC);
+            java.time.Duration duration = java.time.Duration.between(dt.toLocalTime(), queryStartDate.toLocalTime());
+            java.time.Period period;
+
+            if (dt.isAfter(queryStartDate) && duration.compareTo(java.time.Duration.ZERO) > 0) {
```

Comment:
This is tricky code. It *should not* be in a UDF where you can neither test nor debug it. Instead, factor it into a function. Write unit tests for the functions for all the three paths (and other corner cases you can think up.) Call that function from here.

Given the number of function calls in this code, and the number of memory allocations, let's not worry about the overhead of adding one more function layer. Let's instead ensure the cases work.

Reply:
I rewrote the function implementation using an idea from the Joda time BasePeriod